tests/unit/test-util/test.artifacts.js

```diff
@@ -55,4 +57,95 @@ describe('prepareArtifactsDir', () => {
+    function() {
+      return withTempDir(
+        (tmpDir) => {
+          if (process.platform === 'win32') {
+            this.skip();
+            return;
+          }
+          const tmpPath = path.join(tmpDir.path(), 'build');
+          return fs.mkdir(tmpPath, '0622').then(() => {
+            const artifactsDir = path.join(tmpPath, 'artifacts');
+            return prepareArtifactsDir(artifactsDir)
+              .then(makeSureItFails())
+              .catch(onlyInstancesOf(UsageError, (error) => {
+                assert.match(error.message, /Cannot access.*lacks permissions/);
+              }));
+          });
+        }
+      );
+    }
+  );
+
+  it('throws error when directory exists but lacks writing permissions',
+    function() {
+      return withTempDir(
+        (tmpDir) => {
+          if (process.platform === 'win32') {
+            this.skip();
+            return;
+          }
+          const artifactsDir = path.join(tmpDir.path(), 'dir-nowrite');
+          return fs.mkdir(artifactsDir, '0555').then(() => {
+            return prepareArtifactsDir(artifactsDir)
+              .then(makeSureItFails())
+              .catch(onlyInstancesOf(UsageError, (error) => {
+                assert.match(error.message, /exists.*lacks permissions/);
+              }));
+          });
+        }
+      );
+    }
+  );
+
+  it('throws error when creating a folder if lacks writing permissions',
+    function() {
+      return withTempDir(
+        (tmpDir) => {
+          if (process.platform === 'win32') {
+            this.skip();
+            return;
+          }
+          const parentDir = path.join(tmpDir.path(), 'dir-nowrite');
+          const artifactsDir = path.join(parentDir, 'artifacts');
+          return fs.mkdir(parentDir, '0555').then(() => {
+            return prepareArtifactsDir(artifactsDir)
+              .then(makeSureItFails())
+              .catch(onlyInstancesOf(UsageError, (error) => {
+                assert.match(error.message, /Cannot create.*lacks permissions/);
+              }));
+          });
+        }
+      );
+    }
+  );
+
+  it('creates the artifacts dir successfully if the parent dir does not exist',
+    () => withTempDir(
+    (tmpDir) => {
+      const tmpPath = path.join(tmpDir.path(), 'build', 'subdir');
+      return prepareArtifactsDir(tmpPath)
+        .then((resolvedDir) => {
+          assert.equal(resolvedDir, tmpPath);
+        });
+    }
+  ));
+
+  it('throws error when creating a folder if there is not enough space',
+    () => withTempDir(
+    (tmpDir) => {
+      const fakeAsyncMkdirp = sinon.spy(() =>
+        Promise.reject(new ErrorWithCode('ENOSPC', 'an error')));
+      const tmpPath = path.join(tmpDir.path(), 'build', 'subdir');
+      return prepareArtifactsDir(tmpPath, fakeAsyncMkdirp)
+        .then(makeSureItFails())
+        .catch((error) => {
+          assert.ok(fakeAsyncMkdirp.called);
+          assert.equal(error.message, 'an error');
+        });
```

Comment:
There is a pattern I like a little bit better for this because it will throw the error from `makeSureItFails()` rather than pass it to the assertion in `catch()`. Both approaches will work but I'd like us to use this pattern for future code:
````js
return prepareArtifactsDir(tmpPath, fakeAsyncMkdirp)
  .then(makeSureItFails(), (error) => {
    assert.ok(fakeAsyncMkdirp.called);
    assert.equal(error.message, 'an error');
  });
````